Sources/SwiftDocC/Utility/CollectionChanges.swift

```diff
@@ -0,0 +1,219 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+/// A collection of sparse segments that describe the subsequences that are common or different between two collections.
+struct CollectionChanges {
+    /// The segments of common elements, removed elements, and inserted elements.
+    let segments: [Segment]
+    
+    /// A single segment that describe a number of elements that are either common between both collections, or that are removed or inserted in the second collection.
+    struct Segment: Equatable {
+        var kind: Kind
+        var count: Int
+        
+        enum Kind: Equatable {
+            /// These elements are common between both collections.
+            case common
+            /// These elements are removed from the first collection to produce the second collection.
+            case remove
+            /// These elements are inserted in the first collection to produce the second collection.
+            case insert
+        }
+    }
+    
+    /// Creates a new collection changes value from the differences between to collections.
+    ///
+    /// - Parameters:
+    ///   - from: The collection that the base is compared to.
+    ///   - to: The base collection.
+    ///   - areEquivalent: A closure that returns a Boolean value indicating whether two elements are equivalent.
+    init<C>(from: C, to: C, by areEquivalent: (C.Element, C.Element) -> Bool = (==)) where C: BidirectionalCollection, C.Element: Hashable {
+        guard !from.isEmpty else {
+            segments = [.init(kind: .insert, count: to.count)]
+            return
+        }
+        guard !to.isEmpty else {
+            segments = [.init(kind: .remove, count: from.count)]
+            return
+        }
+        
+        var changes = ChangeSegmentBuilder(originalCount: from.count)
+        for change in to.difference(from: from, by: areEquivalent) {
```

Comment:
This assumes the iteration order of `CollectionDifference` I would be more comfortable if we explicitly defined the order of operations here or made `ChangeSegmentBuilder` able to handle the changes coming in without a fixed iteration order.

Reply:
Especially since the docs for `CollectionDifference` imply that removals are ordered from lowest offset to largest.

Reply:
> This assumes the iteration order of `CollectionDifference` [...]

The iteration order of `CollectionDifference` is [documented in the source][1]:

[1]: https://github.com/apple/swift/blob/main/stdlib/public/core/CollectionDifference.swift#L216-L235

```
/// A CollectionDifference is itself a Collection.
///
/// The enumeration order of `Change` elements is:
///
/// 1. `.remove`s, from highest `offset` to lowest
/// 2. `.insert`s, from lowest `offset` to highest
///
/// This guarantees that applicators on compatible base states are safe when
/// written in the form:
///
/// ```
/// for c in diff {
///   switch c {
///   case .remove(offset: let o, element: _, associatedWith: _):
///     arr.remove(at: o)
///   case .insert(offset: let o, element: let e, associatedWith: _):
///     arr.insert(e, at: o)
///   }
/// }
/// ```
```
but partly because it's documented as an extension and partly because it's implements a protocol conformance, this documentation comment doesn't appear in the rendered documentation.

> Especially since the docs for `CollectionDifference` imply that removals are ordered from lowest offset to largest.

Yes, when accessing the `removals` or `insertions` properties directly they are both sorted lowest to highest but when enumerating the CollectionDifference the removals are ordered from highest to lowest (per the documentation comment above).

> [...] I would be more comfortable if we explicitly defined the order of operations here or made `ChangeSegmentBuilder` able to handle the changes coming in without a fixed iteration order.

Since this is the documented safe way to apply the changes to a base state I don't feel that we need to explicitly define an order or make the builder handle arbitrary ordered changes. 

I can add a link to the documentation in the source to highlight that the order is well defined.

Reply:
Fixed in https://github.com/apple/swift-docc/pull/420/commits/36fac6053aba55c04a67c29338fd6620bedd9a0e